### PR TITLE
Feature: split form for outputfile of HS matrix and wfc

### PR DESCRIPTION
**Pull request by lyb9812**

### CODE REVIEW

File: source/module_hamilt_lcao/hamilt_lcaodft/local_orbital_wfc.cpp [hunk 1]
```diff
@@ -153,12 +153,12 @@ int Local_Orbital_wfc::localIndex(int globalindex, int nblk, int nprocs, int& my
...
 
 #ifdef __MPI
-void Local_Orbital_wfc::wfc_2d_to_grid(
-    int out_wfc_lcao, 
-    const double* wfc_2d, 
-    double** wfc_grid, 
-    const ModuleBase::matrix& ekb, 
-    const ModuleBase::matrix& wg)
+void Local_Orbital_wfc::wfc_2d_to_grid(const int istep,
+                                       int out_wfc_lcao,
```
Comment (mohanchen) [inline source/module_hamilt_lcao/hamilt_lcaodft/local_orbital_wfc.cpp]: const int out_wfc_lcao?

File: source/module_hamilt_lcao/hamilt_lcaodft/local_orbital_wfc.cpp [hunk 2]
```diff
@@ -230,15 +238,14 @@ void Local_Orbital_wfc::wfc_2d_to_grid(
...
-void Local_Orbital_wfc::wfc_2d_to_grid(
-    int out_wfc_lcao,
-    const std::complex<double>* wfc_2d,
-    std::complex<double>** wfc_grid,
-    int ik, 
-    const ModuleBase::matrix& ekb, 
-    const ModuleBase::matrix& wg,
-    const std::vector<ModuleBase::Vector3<double>>& kvec_c)
+void Local_Orbital_wfc::wfc_2d_to_grid(const int istep,
+                                       int out_wfc_lcao,
```
Comment (mohanchen) [inline source/module_hamilt_lcao/hamilt_lcaodft/local_orbital_wfc.cpp]: const int out_wfc_lcao?

File: source/module_hamilt_lcao/hamilt_lcaodft/local_orbital_wfc.h [hunk 3]
```diff
@@ -61,12 +61,14 @@ class Local_Orbital_wfc
     // (in which the implementation should be put in header file )
     // because sub-function `write_wfc_nao_complex`contains GlobalC declared in `global.h`
     // which will cause lots of "not defined" if included in a header file.
-    void wfc_2d_to_grid(int out_wfc_lcao,
+    void wfc_2d_to_grid(const int istep,
+                        int out_wfc_lcao,
```
Comment (mohanchen) [inline source/module_hamilt_lcao/hamilt_lcaodft/local_orbital_wfc.h]: const int out_wfc_lcao? also please check other places

**Review by mohanchen — APPROVED** [review]

File: source/module_hamilt_lcao/hamilt_lcaodft/local_orbital_wfc.cpp [hunk 4]
```diff
@@ -299,7 +306,15 @@ void Local_Orbital_wfc::wfc_2d_to_grid(
...
     {
         std::stringstream ss;
-        ss << GlobalV::global_out_dir << "LOWF_K_" << ik + 1 << ".dat";
+        if (GlobalV::out_app_flag)
+        {
+            ss << GlobalV::global_out_dir << "LOWF_K_" << ik + 1 << ".dat";
+        }
+        else
+        {
+            ss << GlobalV::global_out_dir << istep << "_"
```
Comment (hongriTianqi) [inline source/module_hamilt_lcao/hamilt_lcaodft/local_orbital_wfc.cpp:315]: Is it better to write the file name as LOWF_K_1_STEP_1? And please add a CI test for this case.

**Review by hongriTianqi — CHANGES_REQUESTED** [review]

**Comment by hongriTianqi** [comment]
> No description provided.

Is this PR related to any ISSUE?

**Comment by lyb9812** [comment]
> > No description provided.
> 
> Is this PR related to any ISSUE?

Fix #2520

File: source/module_esolver/esolver_ks_lcao.cpp [hunk 5]
```diff
@@ -625,21 +625,23 @@ void ESolver_KS_LCAO::updatepot(const int istep, const int iter)
             }
             bool bit = false; // LiuXh, 2017-03-21
             // if set bit = true, there would be error in soc-multi-core calculation, noted by zhengdy-soc
-            if (this->psi != nullptr)
+            if (this->psi != nullptr && (istep % GlobalV::out_interval == 0))
```
Comment (dyzheng) [inline source/module_esolver/esolver_ks_lcao.cpp:623]: http://abacus.deepmodeling.com/en/latest/advanced/input_files/input-main.html#out-interval the parameter "out_interval" not support this print which described in documentation. 
@lyb9812 please discuss with @YuLiu98  for it.

Comment (lyb9812) [reply source/module_esolver/esolver_ks_lcao.cpp:623]: Ok, I will update the documentation.

File: source/module_hamilt_lcao/hamilt_lcaodft/local_orbital_wfc.cpp (hunk 4 above)
Comment (lyb9812) [reply source/module_hamilt_lcao/hamilt_lcaodft/local_orbital_wfc.cpp:315]: The other matrices are all output in this format. Do I need to change them or maintain them?

File: source/module_esolver/esolver_ks_lcao.cpp (hunk 5 above)
Comment (dyzheng) [reply source/module_esolver/esolver_ks_lcao.cpp:623]: Parameter out_interval only used with "calculation == md" now.
What is your designation here?

File: source/module_hamilt_lcao/hamilt_lcaodft/local_orbital_wfc.cpp (hunk 4 above)
Comment (hongriTianqi) [reply source/module_hamilt_lcao/hamilt_lcaodft/local_orbital_wfc.cpp:315]: OK, just keep it.

File: source/module_esolver/esolver_ks_lcao.cpp (hunk 5 above)
Comment (hongriTianqi) [reply source/module_esolver/esolver_ks_lcao.cpp:623]: I added a function in esolver_ks_lcao.cpp in PR #2613, which might also be useful here.
```
bool ESolver_KS_LCAO::md_skip_out(std::string calculation, int istep, int interval)
 {
     if (calculation == "md")
     {
         if (istep % interval != 0)
         {
             return true;
         }
     }
     return false;
 }
```

Comment (lyb9812) [reply source/module_esolver/esolver_ks_lcao.cpp:623]: Yes,  out_interval is also only used with "calculation == md" for out_mat_hs and out_wfc_lcao.

**Review by hongriTianqi — APPROVED** [review]

**Pull request merged**